py/kubeflow/testing/cd/close_old_prs.py

```diff
@@ -0,0 +1,225 @@
+"""If there are multiple PRs open to update an app; close the older ones"""
+
+import collections
+from dateutil import parser as date_parser
+import fire
+import json
+import logging
+import re
+
+from code_intelligence import github_app
+from code_intelligence import graphql
+
+# The name of the GitHub user under which kubeflow-bot branches exist
+KUBEFLOW_BOT = "kubeflow-bot"
+
+# This needs to be in sync with the pattern used in create_manifests_pr
+# The pattern should be "update_{image.name}_{image.tag}"
+# TODO(jlewi): What if there are underscores in the image name
+HEAD_PATTERN = re.compile("update_([^_]+)_([^_]+)-([^_]+)")
+
+APP_TAG = collections.namedtuple("APP_TAG", ("app", "tag"))
+class PRCloser:
+
+  def __init__(self):
+    self._client = graphql.GraphQLClient()
+
+    self._headers = None
+    self._token_refresher = None
+    # Try various methods to obtain credentials
+    try:
+      self._token_refresher = github_app.FixedAccessTokenGenerator.from_env()
+
+    except ValueError:
+      logging.info("Could not create a FixedAccessTokenGenerator; will try "
+                   "other methods for obtaining credentials")
+
+    if not self._token_refresher:
+      app = github_app.GitHubApp.create_from_env()
+      self._token_refresher = github_app.GitHubAppTokenGenerator(
+        app, "kubeflow/manifests")
+
+
+  def _run_query(self, *args, **kwargs):
+    if not kwargs:
+      kwargs = {}
+    kwargs["headers"] = self._token_refresher.auth_headers
+    return self._client.run_query(*args, **kwargs)
+
+  def apply(self):
+    app_prs = collections.defaultdict(lambda: [])
+
+    for pr in self._iter_prs("kubeflow", "manifests"):
+      login = pr["author"]["login"]
+      if login == KUBEFLOW_BOT:
+        m = HEAD_PATTERN.match(pr["headRefName"])
+
+        if not m:
+          url = pr["url"]
+          logging.error(f"PR: {url} has ref {pr['headRefName']} which doesn't "
+                        f"match pattern {HEAD_PATTERN.pattern}")
+          continue
+
+        app = APP_TAG(m.group(1), m.group(2))
+        app_prs[app] = app_prs[app] + [pr]
+
+    # For each application and tuple close all but the most recent PRs
+    for app, prs in app_prs.items():
+      logging.info(f"{app} has {len(prs)} open prs")
+      if len(prs) == 1:
+        continue
+
+      # We sort the PRs by URL since this will correspond to PR number
+      # We will keep the most recent one.
```

Comment:
What does the PR url looks like?

If URLs are in the format
`https://github.com/kubeflow/testing/pull/99`
`https://github.com/kubeflow/testing/pull/100`

Sorting lexicographically yields unexpected outcome?

Reply:
Good catch. I'll fix this.